tsconfig.base.json

```diff
@@ -0,0 +1,26 @@
+{
+  "compilerOptions": {
+    "lib": ["es2015.iterable", "es2015.generator", "es5"],
+    "moduleResolution": "node",
+
+    "declaration": true,
+    "declarationMap": true,
+    "sourceMap": true,
+    "composite": true,
+    "noEmitOnError": true,
+
+    "strictNullChecks": true,
+    "noImplicitAny": true,
+    "noImplicitThis": true,
+    "strictPropertyInitialization": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+
+    "alwaysStrict": true,
+    "preserveConstEnums": true,
+
+    "jsx": "preserve",
+
+    "types": []
+  }
+}
```

Comment:
I'm not sure this is the PR in which we want to define the overall monorepo's tsconfig file. I'd be more comfortable with each respective package defining their own for now until we have a global baseline that is defined by the TS working group.

Separately, It might be worth generating the basic tsconfig files using the `tsc --init` command since it outputs all *(read: most)* possible settings and gives nice context to each of them. Then from there, we can make changes to the configs above and beyond the defaults as needed.